### ISSUE TITLE
MQTT 5 uTransport topic structure needs to be revisited for d2c & c2d connections

### ISSUE THREAD

**Issue by devkelley** [eclipse-uprotocol/up-spec#192]
This issue was raised in the `uprotocol-mqtt` slack channel:

"a critical flaw in our D2C and C2D topic definitions that there will be too many topics across the millions of vehicles and hundreds of topics per-vehicle if we events are published to the specific topics following: {client_identifier}/{source.authority_name}/{source.ue_id}/{source.ue_version_major}/{source.resource_id}/{sink.authority_name}/{sink.ue_id}/{sink.ue_version_major}/{sink.resource_id}
To simplify the communication (and to drastically reduce the # of topics the broker needed to manage), we used a single D2C and C2D topic that was more like:  {client_identifier}/{sink.authority_name}/ and the sink authority was added to published events sent to the vehicles (so we can ensure all events made it to the vehicles)."

Desired outcome: A MQTT topic structure that allows for authorization controls at the broker level, conforms to the segment limitations of cloud brokers, and reduces the number of unique topics for each vehicle.

Please use this issue to discuss use cases and proposals.

**Comment by stevenhartley** [comment]
Hi @jjj-vtm, What I recall the concern was not about the number of subscriptions as you pointed out we subscribe to a lot of wildcard sections of the topic, but the amount of individual publish topics. I don't understand the problem so I will ask them to comment on this issue now.

**Comment by jjj-vtm** [comment]
Hi @stevenhartley, ok what I can think of and this is a real problem is when for example every ue would instantiate a UTransport (MQTT) on its own and subscribe to the wild card topic. This would lead to many connections and many unnecessary messages. 

Maybe it is already considered in the uProtocol but a Gateway component should solve this issue. For MQTT one would have an internal MQTT broker where each ue subscribes to the very specific topic it wants to listen too and the Gateway would be an UStreamer (MQTT <--> MQTT) which subscribes to the wildcard topic on the cloud broker an forwards all messages to the internal broker. With this topology one would only need a single MQTT connection to the cloud broker and it also has the benefit that one might think of more sophisticated authorization checks on the gateway.

**Comment by stevenhartley** [comment]
That is what we have but we are questioning if the topic structure can be the same across the brokers so that the utransport implementations can remain the same.

**Comment by stevenhartley** [comment]
Below is the information that was placed in #201 for the solution to this issue:

Some MQTT brokers (like Azure) have a limits on the amount of subscriptions and segments for the MQTT topic that would make the current proposed specification non-compatible. Azure MQTT Broker only supports a maximum of 8 segments whereas uProtocol defines 9. Furthermore, Azure also limits the number of subscriptions to 50 that will not work with the current uProtocol specifications that require an MQTT_SUBSCRIBE for every remote topic .

The agreed upon solution was to go back to a smaller and simplified topic definition for D2D communication (between streamers) similar to what we currently use in production that will work with broker implementations like Azure that is the topic:
/{source.authority_name}/{sink.authority_name}

Devices then (when using the transport in D2D mode) MQTT_SUBSCRIBE to anything sent to them +/{my_authority_name} and then MQTT_PUBLISH (send) to the destination topic based on the UUri in the sink (ex. {my_authority_name}/{cloud_authority_name}

Unfortunately the above simplifications won't work for published events as the sink address is not populated inside of published uMessages that are fed to the transport, that is why I was reminded why we manually inserted the sink address inside of published messages before we sent them to the streamer in 1.3.6 implementations in production.

so in summary, to support the above proposal, we have to change the specifications to not invalidate publish uMessages that contain sink (change in uattributes.adoc) and update mqtt5.adoc to support the 2 modes (D-2-D and uE-2-uE).
